# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: js/src/vm/EnvironmentObject.cpp
# Commit: fa99e2e2b957
# Full Hash: fa99e2e2b957d91bf31e2237b2db945d600de169
# Author: Tooru Fujisawa <arai_a@mac.com>
# Date: 2025-07-08 16:25:02
# Regressor Bug: 1970388
# File Overlap Count: 1
# Description:
#   Bug 1970388 - Part 2: Reflect the enclosing environment to MissingEnvironmentMap. r=bthrall,jonco
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D253505
# ==============================================================================

diff -r 90960ec20747 -r fa99e2e2b957 js/src/vm/EnvironmentObject.cpp
--- a/js/src/vm/EnvironmentObject.cpp	Tue Jul 08 11:37:54 2025 +0000
+++ b/js/src/vm/EnvironmentObject.cpp	Tue Jul 08 11:37:54 2025 +0000
@@ -1553,15 +1553,47 @@
   return false;
 }
 
+bool MissingEnvironmentKey::initFromEnvironmentIter(JSContext* cx,
+                                                    const EnvironmentIter& ei) {
+  frame_ = ei.maybeInitialFrame();
+  scope_ = ei.maybeScope();
+  if (frame_) {
+    nearestEnvId_ = 0;
+    return true;
+  }
+
+  EnvironmentObject* env = nullptr;
+  EnvironmentIter copy(cx, ei);
+  while (copy) {
+    if (copy.hasAnyEnvironmentObject()) {
+      env = &copy.environment();
+      break;
+    }
+    ++copy;
+  }
+
+  // The global object should have an environment object even if we don't find
+  // anything else.
+  MOZ_ASSERT(env);
+
+  if (!gc::GetOrCreateUniqueId(env, &nearestEnvId_)) {
+    ReportOutOfMemory(cx);
+    return false;
+  }
+
+  return true;
+}
+
 /* static */
 HashNumber MissingEnvironmentKey::hash(MissingEnvironmentKey ek) {
-  return size_t(ek.frame_.raw()) ^ size_t(ek.scope_);
+  return mozilla::HashGeneric(ek.frame_.raw(), ek.nearestEnvId_, ek.scope_);
 }
 
 /* static */
 bool MissingEnvironmentKey::match(MissingEnvironmentKey ek1,
                                   MissingEnvironmentKey ek2) {
-  return ek1.frame_ == ek2.frame_ && ek1.scope_ == ek2.scope_;
+  return ek1.frame_ == ek2.frame_ && ek1.nearestEnvId_ == ek2.nearestEnvId_ &&
+         ek1.scope_ == ek2.scope_;
 }
 
 bool LiveEnvironmentVal::traceWeak(JSTracer* trc) {
@@ -2789,21 +2821,29 @@
 }
 
 /* static */
-DebugEnvironmentProxy* DebugEnvironments::hasDebugEnvironment(
-    JSContext* cx, const EnvironmentIter& ei) {
+bool DebugEnvironments::getExistingDebugEnvironment(
+    JSContext* cx, const EnvironmentIter& ei, DebugEnvironmentProxy** out) {
   MOZ_ASSERT(!ei.hasSyntacticEnvironment());
 
   DebugEnvironments* envs = cx->realm()->debugEnvs();
   if (!envs) {
-    return nullptr;
-  }
-
-  if (MissingEnvironmentMap::Ptr p =
-          envs->missingEnvs.lookup(MissingEnvironmentKey(ei))) {
+    *out = nullptr;
+    return true;
+  }
+
+  MissingEnvironmentKey key;
+  if (!key.initFromEnvironmentIter(cx, ei)) {
+    return false;
+  }
+
+  if (MissingEnvironmentMap::Ptr p = envs->missingEnvs.lookup(key)) {
     MOZ_ASSERT(CanUseDebugEnvironmentMaps(cx));
-    return p->value();
-  }
-  return nullptr;
+    *out = p->value();
+    return true;
+  }
+
+  *out = nullptr;
+  return true;
 }
 
 /* static */
@@ -2822,7 +2862,10 @@
     return false;
   }
 
-  MissingEnvironmentKey key(ei);
+  MissingEnvironmentKey key;
+  if (!key.initFromEnvironmentIter(cx, ei)) {
+    return false;
+  }
   MOZ_ASSERT(!envs->missingEnvs.has(key));
   if (!envs->missingEnvs.put(key,
                              WeakHeapPtr<DebugEnvironmentProxy*>(debugEnv))) {
@@ -3029,9 +3072,16 @@
   MOZ_ASSERT(ei.withinInitialFrame());
   MOZ_ASSERT(ei.scope().is<Scope>());
 
+  MissingEnvironmentKey key;
+  {
+    js::AutoEnterOOMUnsafeRegion oomUnsafe;
+    if (!key.initFromEnvironmentIter(cx, ei)) {
+      oomUnsafe.crash("OOM during onPopGeneric");
+      return;
+    }
+  }
   Rooted<Environment*> env(cx);
-  if (MissingEnvironmentMap::Ptr p =
-          envs->missingEnvs.lookup(MissingEnvironmentKey(ei))) {
+  if (MissingEnvironmentMap::Ptr p = envs->missingEnvs.lookup(key)) {
     env = &p->value()->environment().as<Environment>();
     envs->missingEnvs.remove(p);
   } else if (ei.hasSyntacticEnvironment()) {
@@ -3264,9 +3314,12 @@
               ei.scope().is<WasmFunctionScope>() || ei.scope().is<VarScope>() ||
               ei.scope().kind() == ScopeKind::StrictEval));
 
-  if (DebugEnvironmentProxy* debugEnv =
-          DebugEnvironments::hasDebugEnvironment(cx, ei)) {
-    return debugEnv;
+  DebugEnvironmentProxy* maybeDebugEnv;
+  if (!DebugEnvironments::getExistingDebugEnvironment(cx, ei, &maybeDebugEnv)) {
+    return nullptr;
+  }
+  if (maybeDebugEnv) {
+    return maybeDebugEnv;
   }
 
   EnvironmentIter copy(cx, ei);